algebra2/gaussian_elimination/gaussian_elimination.h

```diff
@@ -0,0 +1,566 @@
+            for(int j = i+1; j < m.extent(0); j++){
+                if(m[j][i].numerator != 0){
+                    swap(m, i, j);
+                }
+            }
+        }
+        if(m[i][i].numerator != 0){
+            for(int k = i+1; k < m.extent(0); k++){
+                res = m[k][i]/m[i][i];
+                if(res.numerator != 0){
+                    subtract(m, k, i, res);
+                }
+            }
+        }
+    }
+
+    if(reducted == algorithms::gaussian_elimination::reducted_form::diagonal){
+        for(int i = 0;  i < m.extent(0); i++){
+            if(m[i, i].numerator != 0){
+                for(int k = i+1; k < m.extent(0); k++){
+                    res = m[i, k]/m[k, k];
+                    if(res.numerator != 0){
+                        subtract(m, i, k, res);
+                    }
+                }
+            }
+        }
+    }
+
+    std::println("The reduced matrix is {}\n", m);
+}
+
+template<typename T, typename LP>
+auto run(ranges::matrix_view<T, LP> m,
+         algorithms::gaussian_elimination::axis a,
+         algorithms::gaussian_elimination::operation allowed_operations = algorithms::gaussian_elimination::operation::swap | algorithms::gaussian_elimination::operation::add |
+                                                                          algorithms::gaussian_elimination::operation::multiply,
+         algorithms::gaussian_elimination::reducted_form reducted = algorithms::gaussian_elimination::reducted_form::echelon) -> void {
+
+    std::println("//////////Gaussian Elimination on matrix m//////////\n");
+    std::println("Matrix m = {}\n", m);
+    double res;
+
+    for(int i = 0;  i < m.extent(0); i++) {
+        if (m[i, i] == 0) {
+            for (int j = i + 1; j < m.extent(0); j++) {
+                if (m[j, i] != 0) {
+                    swap(m, i, j);
+                    break;
+                }
+            }
+        }
+        if (m[i, i] != 0) {
+            for (int k = i + 1; k < m.extent(0); k++) {
+                res = m[k, i] / m[i, i];
+                if (res != 0) {
+                    subtract(m, k, i, res);
+                }
+            }
+        }
+    }
+
+    if(reducted == algorithms::gaussian_elimination::reducted_form::diagonal){
+        for(int i = 0;  i < m.extent(0); i++){
+            if(m[i, i] != 0){
+                for(int k = i+1; k < m.extent(0); k++){
+                    res = m[i, k]/m[k, k];
+                    if(res != 0){
+                        subtract(m, i, k, res);
+                    }
+                }
+            }
+        }
+    }
+    std::println("The reduced matrix is {}\n", m);
+}
+
+
+/*
+ * description:
+ *      Performs Gaussian elimination on matrix m containing fractions and shows steps after each operation*/
+template<std::integral T, typename LP>
+auto show_steps(ranges::matrix_view<ranges::fraction<T>, LP> m,
+                algorithms::gaussian_elimination::axis a,
```

Comment:
once more rep of code - see e.g. gcd implementationof your classmates to see how to avoid such a situation. Moreover, you should not print anything - return a std::string containing description.